### PR TITLE
Add cw4 hooks

### LINKED ISSUE #140: Add listeners to cw4-group (and cw4?)
We should allow multiple contracts to be informed of changing group membership. This is a key feature so we can remove the restriction of immutible groups from #80 

- [x] Define a message format the listeners receive (a diff of changes to weights)
- [x] Define a message to "register a listener". Note that if I register a listener that always errors, it will prevent you from ever updating the group - only the group owner should be able to add a listener (same as change members)
- [x] After [`UpdateMembers`](https://github.com/CosmWasm/cosmwasm-plus/blob/master/packages/cw4/src/msg.rs#L31-L34) is executed, all registered listeners will get sent a message of this update in the same transaction.

Please code and test this in cw4-group, and update the cw4 spec to include this (both the types and the README)

Note, in some cases like #141 we may want to know both before and after weight for the changes, not just the new state. Consider how much overhead that information adds (do we need to make more queries? or do we make those already?) and what benefit it provides. You do not need to implement it, but please address this in the comments of the PR

### PR DESCRIPTION
**Pull request by ethanfrey**
Closes #140 

- [x] Update cw4 spec
- [x] Update cw4-group implementation
- [x] Test implementation
- [x] Update READMEs

### CODE REVIEW

File: contracts/cw4-group/src/contract.rs [hunk 1]
```diff
@@ -92,31 +104,37 @@ pub fn update_members(
...
     let mut total = TOTAL.load(deps.storage)?;
+    let mut diffs: Vec<MemberDiff> = vec![];
 
     // add all new members and update total
     for add in to_add.into_iter() {
         let raw = deps.api.canonical_address(&add.addr)?;
         MEMBERS.update(deps.storage, &raw, |old| -> StdResult<_> {
             total -= old.unwrap_or_default();
             total += add.weight;
+            diffs.push(MemberDiff::new(add.addr, old, Some(add.weight)));
```
Comment (maurolacy) [inline contracts/cw4-group/src/contract.rs:120]: Very nice.

File: contracts/cw4-group/src/contract.rs [hunk 2]
```diff
@@ -92,31 +104,37 @@ pub fn update_members(
...
             Ok(add.weight)
         })?;
     }
 
     for remove in to_remove.into_iter() {
         let raw = deps.api.canonical_address(&remove)?;
-        total -= MEMBERS.may_load(deps.storage, &raw)?.unwrap_or_default();
-        MEMBERS.remove(deps.storage, &raw);
+        let old = MEMBERS.may_load(deps.storage, &raw)?;
+        // only process this if they were actually in the lsit before
```
Comment (maurolacy) [inline contracts/cw4-group/src/contract.rs:128]:
```suggestion
        // Only process this if they were actually in the list before
```

File: contracts/cw4-group/src/contract.rs [hunk 3]
```diff
@@ -380,6 +440,170 @@ mod tests {
...
+        // same order as in the message (adds first, then remove)
+        let diffs = vec![
+            MemberDiff::new(USER1, Some(11), Some(20)),
+            MemberDiff::new(USER3, None, Some(5)),
+            MemberDiff::new(USER2, Some(6), None),
+        ];
+        let hook_msg = MemberChangedHookMsg { diffs };
+        let msg1 = hook_msg.clone().into_cosmos_msg(contract1).unwrap();
+        let msg2 = hook_msg.into_cosmos_msg(contract2).unwrap();
+        assert_eq!(res.messages, vec![msg1, msg2]);
```
Comment (maurolacy) [inline contracts/cw4-group/src/contract.rs:604]: Nice check.

**Review by maurolacy — APPROVED** [review]
lgtm.

File: packages/cw4/src/hook.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            old_weight,
+            new_weight,
+        }
+    }
+}
+
+impl MemberChangedHookMsg {
+    /// serializes the message
+    pub fn into_binary(self) -> StdResult<Binary> {
+        let msg = MemberChangedHandleMsg::MemberChangedHook(self);
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply packages/cw4/src/hook.rs:42]: Yes, it must be embedded one layer down. Very unlikely that any contract uses `MemberChangedHandleMsg`, but they must handle a superset of that.

File: contracts/cw4-group/src/contract.rs [hunk 5]
```diff
@@ -77,13 +82,20 @@ pub fn update_admin(
...
-    update_members(deps, info.sender, add, remove)?;
-    Ok(HandleResponse::default())
+    // make the local update
+    let diff = update_members(deps.branch(), info.sender, add, remove)?;
+    // call all registered hooks
+    let mut ctx = Context::new();
+    for h in HOOKS.may_load(deps.storage)?.unwrap_or_default() {
+        let msg = diff.clone().into_cosmos_msg(h)?;
+        ctx.add_message(msg);
+    }
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply contracts/cw4-group/src/contract.rs:97]: Exactly, the Context was made to make that look a bit nicer.
Jury still a bit out if it is better, but happy for opinions.

[the remark this answers is not in the document]
Comment (maurolacy) [reply contracts/cw4-group/src/contract.rs:97]: Maybe it's because I'm just used to it, but I prefer the other variant better.

**Pull request merged**